hw_x86_64 changes

Kensan opened the pull request:
This set of changes contains some fixes and one extension for hw_x86_64:

- The initial page tables have been extracted so they can be changed easily for platforms based on x86_64.
- Several log messages have been extended to provide the same amount of debug information as on ARM platforms.
- FPU initialization and lazy-state handling has been fixed to work properly on hardware.
- Write entire 64 bits of the I/O APIC interrupt redirection table entries when initializing the x86 PIC.
- Populate the I/O memory allocator with the entire address space excluding the core only RAM regions, aligning the MMIO allocator initialization with how it is done for other x86 kernels.

The last item removes the limitation discussed in the last comment of #1448.

m-stein commented:
@Kensan 
Cool that you had the time to enhance the X86 support!
I've added some style fixes. Do you agree with them? Could you please have a look especially at my modifications in ``pic.h``. They have some potential to introduce bugs.

Kensan commented:
@m-stein:
We revisited the hw_x86_64 code and did tests with real hardware to have a solid basis for porting base-hw to the Muen Separation Kernel.

The style fixes are fine from my side. I will verify the other changes and get back to you.

m-stein commented:
@Kensan 
Thank you. One thing I wasn't sure about is whether an IRTE can be treated as 64 bit value consistently (except when writing to IOWIN) because you initially declared it as uint32_t to cast it to uint64_t later.

Kensan commented:
I reviewed the modifications and they are great. I updated the branch and added a minor correction to the I/O mem allocator comment on top.

Thanks for taking the time to review our changes.

chelmuth commented:
I rebased the commits on master and pushed 127f5d34afe46eafc89f86d3837d6c5a04c77919 to my fork. @m-stein and @skalk what are your opinions about merging the commits to staging?

chelmuth commented:
Sorry I mixed up the codelabs-ch branches. The proper tree is df8688eb89eeae0765a214f3c44771b9cfb5cf59. Thanks @Kensan for the hint.

chelmuth commented:
Our nightly tests went rather good with the changes, so, you may go for submitting the Muen-related branch too. I have some minor issues for discussions, which are not critical for the merge (moved from IRC to here):

* I'm testing hw_x86_64-changes currently and get some "Unknown interrupt <nr>" messages, e.g., when hitting a key before the PS/2 driver comes up (or is missing from the scenario). Do you ensure that unused IRQs are masked on boot?
* Further I tried to start our acpi_drv on hw_x86_64 but ended up with "I/O memory [fffe000,ffff000) not available". Do you have any hints or is ACPI table usage (e.g., for remapped IRQs) no supported currently?
* Last, our AHCI driver also chokes as hw_x86_64 assumes the AHCI interrupt to be edge-triggered. Any ideas in this regard?

reet- commented:
@chelmuth
Great that most of the tests work!

Considering the Muen-related branch: We have to wait until AdaCore releases the GPL 2015 version of GNAT/SPARK in order to prove/build the latest Muen kernel (and make the steps in the tutorial doc `base-hw/doc/x86_64_muen.txt` work). The release should happen very soon.

* Unknown interrupts: We are not explicitly masking any interrupts on I/O APIC init as this was previously done explicitly in the Irq() ctor. Maybe this has changed now. Will look into it.
* How can I reproduce the ACPI/AHCI driver issues?

chelmuth commented:
For AHCI just try `make run/libc_ffat`.
